lightly/utils/dist.py

```diff
@@ -8,22 +8,21 @@ class GatherLayer(torch.autograd.Function):
     """Gather tensors from all processes, supporting backward propagation.
 
     This code was taken and adapted from here:
-    https://github.com/Spijkervet/SimCLR
+    https://github.com/vturrisi/solo-learn/blob/b69b4bd27472593919956d9ac58902a301537a4d/solo/utils/misc.py#L187
 
     """
 
     @staticmethod
     def forward(ctx, input: torch.Tensor) -> Tuple[torch.Tensor, ...]:
-        ctx.save_for_backward(input)
         output = [torch.empty_like(input) for _ in range(dist.get_world_size())]
         dist.all_gather(output, input)
         return tuple(output)
 
     @staticmethod
-    def backward(ctx, *grads: torch.Tensor) -> torch.Tensor:
-        (input,) = ctx.saved_tensors
-        grad_out = torch.empty_like(input)
-        grad_out[:] = grads[dist.get_rank()]
+    def backward(ctx, *grads) -> torch.Tensor:
+        all_gradients = torch.stack(grads)
```

Comment:
The code here can be optimized by avoiding stacking (see https://github.com/cellarium-ai/cellarium-ml/blob/220ba90b47378c99d4c08b9d91c5c31b796cb3ca/cellarium/ml/distributed/gather.py#L17-L21):

```py
    @staticmethod
    def backward(ctx, *grads) -> torch.Tensor:
        grad_out = grads[dist.get_rank()].contiguous()
        dist.all_reduce(grad_out, op=dist.ReduceOp.SUM)
        return grad_out
```

Reply:
The code you proposed causes the test to fail. I have tested it using two devices and by calling `TestGatherLayer().test(None)`. Thus I would rather keep the existing code, as it is the only one causing the training to come to exactly the same values.

### the code you proposed:
```
grads: (tensor([[-0.0576, -0.0557, -0.0831, -0.0729],
        [-0.0867, -0.0694, -0.0778, -0.0508],
        [-0.0852, -0.0687, -0.0781, -0.0520],
        [ 0.0289, -0.0130, -0.0879, -0.1265]]), tensor([[0.0354, 0.0653, 0.0552, 0.0957],
        [0.0354, 0.0651, 0.0548, 0.0950],
        [0.0353, 0.0649, 0.0546, 0.0946],
        [0.0352, 0.0647, 0.0542, 0.0939]]))
grads: (tensor([[0.0454, 0.0753, 0.0521, 0.0859],
        [0.0457, 0.0759, 0.0524, 0.0865],
        [0.0457, 0.0759, 0.0524, 0.0865],
        [0.0393, 0.0649, 0.0451, 0.0740]]), tensor([[-0.0629, -0.0525, -0.0833, -0.0704],
        [-0.0692, -0.0552, -0.0817, -0.0648],
        [-0.0726, -0.0567, -0.0807, -0.0617],
        [-0.0773, -0.0587, -0.0794, -0.0574]]))
grad_out: tensor([[-0.1205, -0.1082, -0.1664, -0.1434],
        [-0.1559, -0.1246, -0.1595, -0.1156],
        [-0.1579, -0.1254, -0.1588, -0.1136],
        [-0.0484, -0.0718, -0.1674, -0.1839]])
grad_out: tensor([[-0.1205, -0.1082, -0.1664, -0.1434],
        [-0.1559, -0.1246, -0.1595, -0.1156],
        [-0.1579, -0.1254, -0.1588, -0.1136],
        [-0.0484, -0.0718, -0.1674, -0.1839]])
```

### the current code
```
grads: (tensor([[-0.0576, -0.0557, -0.0831, -0.0729],
        [-0.0867, -0.0694, -0.0778, -0.0508],
        [-0.0852, -0.0687, -0.0781, -0.0520],
        [ 0.0289, -0.0130, -0.0879, -0.1265]]), tensor([[0.0354, 0.0653, 0.0552, 0.0957],
        [0.0354, 0.0651, 0.0548, 0.0950],
        [0.0353, 0.0649, 0.0546, 0.0946],
        [0.0352, 0.0647, 0.0542, 0.0939]]))
grads: (tensor([[0.0454, 0.0753, 0.0521, 0.0859],
        [0.0457, 0.0759, 0.0524, 0.0865],
        [0.0457, 0.0759, 0.0524, 0.0865],
        [0.0393, 0.0649, 0.0451, 0.0740]]), tensor([[-0.0629, -0.0525, -0.0833, -0.0704],
        [-0.0692, -0.0552, -0.0817, -0.0648],
        [-0.0726, -0.0567, -0.0807, -0.0617],
        [-0.0773, -0.0587, -0.0794, -0.0574]]))
all_gradients after reduce: tensor([[[-0.0121,  0.0196, -0.0311,  0.0130],
         [-0.0410,  0.0065, -0.0254,  0.0357],
         [-0.0395,  0.0072, -0.0257,  0.0346],
         [ 0.0682,  0.0518, -0.0428, -0.0525]],

        [[-0.0275,  0.0129, -0.0281,  0.0253],
         [-0.0339,  0.0099, -0.0268,  0.0303],
         [-0.0373,  0.0083, -0.0261,  0.0329],
         [-0.0421,  0.0059, -0.0252,  0.0365]]])
all_gradients after reduce: tensor([[[-0.0121,  0.0196, -0.0311,  0.0130],
         [-0.0410,  0.0065, -0.0254,  0.0357],
         [-0.0395,  0.0072, -0.0257,  0.0346],
         [ 0.0682,  0.0518, -0.0428, -0.0525]],

        [[-0.0275,  0.0129, -0.0281,  0.0253],
         [-0.0339,  0.0099, -0.0268,  0.0303],
         [-0.0373,  0.0083, -0.0261,  0.0329],
         [-0.0421,  0.0059, -0.0252,  0.0365]]])
grad_out: tensor([[-0.0121,  0.0196, -0.0311,  0.0130],
        [-0.0410,  0.0065, -0.0254,  0.0357],
        [-0.0395,  0.0072, -0.0257,  0.0346],
        [ 0.0682,  0.0518, -0.0428, -0.0525]])
grad_out: tensor([[-0.0275,  0.0129, -0.0281,  0.0253],
        [-0.0339,  0.0099, -0.0268,  0.0303],
        [-0.0373,  0.0083, -0.0261,  0.0329],
        [-0.0421,  0.0059, -0.0252,  0.0365]])
```

Reply:
You are right, thanks! Each gradient in grads needs to be reduced across all ranks.

I initially had the code below that was working but then switched to the one above without properly testing it with the model.

```py
    @staticmethod
    def backward(ctx, *grads) -> torch.Tensor:
        new_grads = []
        for grad in grads:
            grad = grad.contiguous()
            dist.all_reduce(grad, op=dist.ReduceOp.SUM)
            new_grads.append(grad)
        grad_out = new_grads[dist.get_rank()]
        return grad_out
```

Reply:
Both of the methods you proposed @ordabayevy give the same results:
```
    def backward(ctx, *grads) -> torch.Tensor: 
        new_grads = []
        for grad in grads:
            grad = grad.contiguous()
            dist.all_reduce(grad, op=dist.ReduceOp.SUM)
            new_grads.append(grad)
        print(f"rank: {dist.get_rank()}, new_grads: {new_grads}")
        grad_out = new_grads[dist.get_rank()]
        print(f"rank: {dist.get_rank()}, grad_out: {grad_out}")
        return grad_out
```
```
rank: 0, new_grads: [tensor([[-0.0121,  0.0196, -0.0311,  0.0130],
        [-0.0410,  0.0065, -0.0254,  0.0357],
        [-0.0395,  0.0072, -0.0257,  0.0346],
        [ 0.0682,  0.0518, -0.0428, -0.0525]]), tensor([[-0.0275,  0.0129, -0.0281,  0.0253],
        [-0.0339,  0.0099, -0.0268,  0.0303],
        [-0.0373,  0.0083, -0.0261,  0.0329],
        [-0.0421,  0.0059, -0.0252,  0.0365]])]rank: 1, new_grads: [tensor([[-0.0121,  0.0196, -0.0311,  0.0130],
        [-0.0410,  0.0065, -0.0254,  0.0357],
        [-0.0395,  0.0072, -0.0257,  0.0346],
        [ 0.0682,  0.0518, -0.0428, -0.0525]]), tensor([[-0.0275,  0.0129, -0.0281,  0.0253],
        [-0.0339,  0.0099, -0.0268,  0.0303],
        [-0.0373,  0.0083, -0.0261,  0.0329],
        [-0.0421,  0.0059, -0.0252,  0.0365]])]

rank: 1, grad_out: tensor([[-0.0275,  0.0129, -0.0281,  0.0253],
        [-0.0339,  0.0099, -0.0268,  0.0303],
        [-0.0373,  0.0083, -0.0261,  0.0329],
        [-0.0421,  0.0059, -0.0252,  0.0365]])
rank: 0, grad_out: tensor([[-0.0121,  0.0196, -0.0311,  0.0130],
        [-0.0410,  0.0065, -0.0254,  0.0357],
        [-0.0395,  0.0072, -0.0257,  0.0346],
        [ 0.0682,  0.0518, -0.0428, -0.0525]])
```

Reply:
Resolving this to allow merging